File: app/src/main/java/com/robifr/ledger/ui/create_queue/CreateQueueDate.java
/**
 * Copyright (c) 2022-present Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.create_queue;

import android.view.View;
import androidx.annotation.NonNull;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.robifr.ledger.R;
import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Objects;

public class CreateQueueDate implements View.OnClickListener {
  @NonNull private final CreateQueueFragment _fragment;

  public CreateQueueDate(@NonNull CreateQueueFragment fragment) {
    this._fragment = Objects.requireNonNull(fragment);

    this._fragment.fragmentBinding().date.setOnClickListener(this);
  }

  @Override
  public void onClick(@NonNull View view) {
    Objects.requireNonNull(view);

    final ZonedDateTime inputtedDate =
        this._fragment.createQueueViewModel().inputtedDate().getValue();

    switch (view.getId()) {
      case R.id.date -> {
        final MaterialDatePicker.Builder<Long> pickerBuilder =
            MaterialDatePicker.Builder.datePicker()
                .setTitleText("Select date")
                .setTheme(R.style.MaterialDatePicker)
                .setInputMode(MaterialDatePicker.INPUT_MODE_CALENDAR);

        if (inputtedDate != null) {
          pickerBuilder.setSelection(
              inputtedDate
                  .toLocalDate()
                  .atStartOfDay()
                  .atZone(ZoneId.of("UTC")) // Material only accept UTC time.
                  .toInstant()
                  .toEpochMilli());
        }

        final MaterialDatePicker<Long> picker = pickerBuilder.build();
        picker.show(this._fragment.getChildFragmentManager(), CreateQueueDate.class.toString());
        picker.addOnPositiveButtonClickListener(
            date -> {
              final ZonedDateTime parsedDate =
                  Instant.ofEpochMilli(date).atZone(ZoneId.systemDefault());
              this._fragment.createQueueViewModel().onDateChanged(parsedDate);
            });
      }
    }
  }

  public void setInputtedDate(@NonNull ZonedDateTime date) {
    Objects.requireNonNull(date);

    final String formatted = DateTimeFormatter.ofPattern("EEEE, d MMMM yyyy").format(date);
    this._fragment.fragmentBinding().date.setText(formatted);
  }
}

File: app/src/main/java/com/robifr/ledger/ui/createqueue/CreateQueueDate.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createqueue

import com.google.android.material.R as MaterialR
import com.google.android.material.datepicker.CalendarConstraints
import com.google.android.material.datepicker.DateValidatorPointBackward
import com.google.android.material.datepicker.MaterialDatePicker
import com.robifr.ledger.R
import com.robifr.ledger.util.ClassPath
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

class CreateQueueDate(private val _fragment: CreateQueueFragment) {
  private val _dialog: MaterialDatePicker<Long> =
      MaterialDatePicker.Builder.datePicker()
          .setTheme(MaterialR.style.ThemeOverlay_Material3_MaterialCalendar)
          .setTitleText(R.string.createQueue_date_selectDate)
          .setInputMode(MaterialDatePicker.INPUT_MODE_CALENDAR)
          .setCalendarConstraints(
              CalendarConstraints.Builder().setValidator(DateValidatorPointBackward.now()).build())
          .setSelection(
              _fragment.createQueueViewModel.uiState.safeValue.date
                  .toLocalDate()
                  .atStartOfDay()
                  .atZone(ZoneId.of("UTC")) // Material only accept UTC time.
                  .toInstant()
                  .toEpochMilli())
          .build()
          .apply {
            addOnPositiveButtonClickListener { date: Long? ->
              date?.let {
                _fragment.createQueueViewModel.onDateChanged(
                    Instant.ofEpochMilli(it).atZone(ZoneId.systemDefault()))
              }
            }
          }

  init {
    _fragment.fragmentBinding.date.setOnClickListener {
      _dialog.show(
          _fragment.requireActivity().supportFragmentManager,
          ClassPath.simpleName(CreateQueueDate::class.java))
    }
  }

  fun setInputtedDate(date: ZonedDateTime) {
    _fragment.fragmentBinding.date.setText(
        DateTimeFormatter.ofPattern("EEEE, d MMMM yyyy").format(date))
  }
}